p2pserver/net/netserver/nbr_peers.go

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (C) 2018 The ontology Authors
+ * This file is part of The ontology library.
+ *
+ * The ontology is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ontology is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package netserver
+
+import (
+	"net"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ontio/ontology/p2pserver/peer"
+
+	comm "github.com/ontio/ontology/common"
+	"github.com/ontio/ontology/common/log"
+	"github.com/ontio/ontology/p2pserver/common"
+	"github.com/ontio/ontology/p2pserver/message/types"
+)
+
+// Conn is a net.Conn wrapper to do some clean up when Close.
+type Conn struct {
+	net.Conn
+	session   uint64
+	id        common.PeerId
+	closed    bool
+	netServer *NetServer
+}
+
+// Close overwrite net.Conn
+func (self *Conn) Close() error {
+	self.netServer.Np.Lock()
+	defer self.netServer.Np.Unlock()
+	if self.closed {
+		return nil
+	}
+
+	n := self.netServer.Np.List[self.id]
+	if n.Peer == nil {
+		log.Fatalf("connection %s not in net server", self.id.ToHexString())
+	} else if n.session == self.session { // connection not replaced
+		delete(self.netServer.Np.List, self.id)
+		// need handle asynchronously since we hold Np.Lock
+		log.Infof("remove peer %s from net server", self.id.ToHexString())
+		go self.netServer.notifyPeerDisconnected(n.Peer.Info)
+	}
+
+	self.closed = true
+	return self.Conn.Close()
+}
+
+type connectedPeer struct {
+	session uint64
+	Peer    *peer.Peer
+}
+
+//NbrPeers: The neigbor list
+type NbrPeers struct {
+	sync.RWMutex
+	List map[common.PeerId]connectedPeer
+
+	nextSessionId uint64
+}
+
+func (self *NbrPeers) getSessionId() uint64 {
+	return atomic.AddUint64(&self.nextSessionId, 1)
+}
+
+func NewNbrPeers() *NbrPeers {
+	return &NbrPeers{
+		List: make(map[common.PeerId]connectedPeer),
+	}
+}
+
+//Broadcast tranfer msg buffer to all establish Peer
+func (this *NbrPeers) Broadcast(msg types.Message) {
+	sink := comm.NewZeroCopySink(nil)
+	types.WriteMessage(sink, msg)
+
+	this.RLock()
+	defer this.RUnlock()
```

Comment:
broadcast with RLock?
what happened if sendraw failed?

Reply:
replaced with `go SendRaw` to avoid malicious peer blocking other peers,  can use `SendAsync` when it is landed.